### PR TITLE
Add: retry at acquire lock error, change the shape of the error object

### LINKED ISSUE #19: Use a better error message for acquireLock errors
https://github.com/openstyles/stylus/issues/1201

### PR DESCRIPTION
**Pull request by eight04**
Fixes #19

### CODE REVIEW

**Comment by eight04** [comment]
@tophf WDYT

**Comment by tophf** [comment]
I would add small randomness:
```js
const fuzz = .25;
wait *= retryExp * (1 - fuzz / 2 + Math.random() * fuzz);
```

File: lib/db-to-cloud.js [hunk 1]
```diff
@@ -304,7 +278,27 @@ function dbToCloud({
...
+      try {
+        await drive.acquireLock(lockExpire);
+        break;
+      } catch (err) {
+        if (err.code !== "EEXIST") {
+          throw err;
+        }
+      }
+      
+      wait *= retryExp;
```
Comment (eight04) [inline lib/db-to-cloud.js:293]: This line should be moved to 301

**Comment by eight04** [comment]
> small randomness

It seems that it doesn't make much difference. Bench script:
```js
const time = [];
const attempts = [];
const SIZE = 500000;

for (let i = 0; i < SIZE; i++) {
  test();
}

console.log(`Success rate: ${((time.length / SIZE) * 100).toFixed(2)}%`);
console.log(`Average retried: ${(sumArr(attempts) / attempts.length).toFixed(2)}`);
console.log(`Average time: ${(sumArr(time) / time.length).toFixed(2)}`);

function sumArr(arr) {
  let total = 0;
  for (const i of arr) {
    total += i;
  }
  return total;
}

function test() {
  // suppose collision resolves after 3~60 seconds
  const collision = Math.random() * 57 + 3;
  
  let wait = 10;
  let tried = 0;
  let currTime = 0;
  const fuzz = .25;
  while (true) {
    if (currTime > collision) {
      time.push(currTime);
      attempts.push(tried);
      return;
    }
    currTime += wait;
    // wait *= 1.5 * (1 + fuzz * (Math.random() - 0.5));
    wait *= 1.5;
    tried++;
    if (tried >= 5) {
      return;
    }
  }
}
```

**Comment by tophf** [comment]
`Average retried` in your test becomes noticeably lower: from 2.73 to 2.36. The bigger `collision` is, the bigger is the improvement of this delta. In real life it can be even more than that. I just don't see the point of using exact intervals.

**Comment by eight04** [comment]
> `Average retried` in your test becomes noticeably lower: from 2.73 to 2.36

Interesting. If `Average retried` gets smaller then the delay for each try is getting longer. However, even if I replace `Math.random()` with `1`, I still get `Average retried: 2.54`.

> using exact intervals.

To use an exact interval, you can set `retryExp` to `1`. Though I think the current default (result in `10`, `15`, `22.5`, `33.75`, `50.625` delays) is good enough?

----

One thing I'm thinking is whether it should report the retry progress e.g.
```js
onProgress({
  phase: 'acquireLock'
})
onProgress({
  phase: 'acquireLockError',
  err,
  retryAt: Timestamp
})
```

**Comment by tophf** [comment]
By "exact intervals" I meant "invariant intervals" i.e. 10, 15, 22.5, 33.75, 50.625 that may coincide between two machines that started to sync at the same time. Randomness (fuzziness) solves this problem.

**Pull request merged**